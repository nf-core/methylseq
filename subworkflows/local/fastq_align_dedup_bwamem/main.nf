include { BAM_SORT_STATS_SAMTOOLS                           } from '../../nf-core/bam_sort_stats_samtools/main'
include { FASTQ_ALIGN_BWA                                   } from '../../nf-core/fastq_align_bwa/main'
include { PICARD_ADDORREPLACEREADGROUPS                     } from '../../../modules/nf-core/picard/addorreplacereadgroups/main'
include { PICARD_MARKDUPLICATES as PICARD_REMOVEDUPLICATES  } from '../../../modules/nf-core/picard/markduplicates/main'  
include { PARABRICKS_FQ2BAM                                 } from '../../../modules/nf-core/parabricks/fq2bam/main'
include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_DEDUPLICATED     } from '../../../modules/nf-core/samtools/index/main'

workflow FASTQ_ALIGN_DEDUP_BWAMEM {

    take:
    ch_reads             // channel: [ val(meta), [ reads ] ]
    ch_fasta             // channel: [ val(meta), [ fasta ] ]
    ch_fasta_index       // channel: [ val(meta), [ fasta index ] ]
    ch_bwamem_index      // channel: [ val(meta), [ bwam index ] ]
    skip_deduplication   // boolean: whether to deduplicate alignments
    use_gpu              // boolean: whether to use GPU or CPU for bwamem alignment

    main:

    ch_alignment                     = Channel.empty()
    ch_alignment_index               = Channel.empty()
    ch_flagstat                      = Channel.empty()
    ch_stats                         = Channel.empty()
    ch_picard_metrics                = Channel.empty()
    ch_multiqc_files                 = Channel.empty()
    ch_versions                      = Channel.empty()

    if (use_gpu) {
        /*
        * Align with parabricks GPU enabled fq2bammeth implementation of bwameth
        */
        PARABRICKS_FQ2BAM (
            ch_reads,
            ch_fasta,
            ch_bwamem_index,
            [], // interval file
            [], // known sites
            'bam' // output format
        )
        ch_alignment = PARABRICKS_FQ2BAM.out.bam
        ch_versions  = ch_versions.mix(PARABRICKS_FQ2BAM.out.versions)
        
        // FQ2BAM can also sort and markduplicates
        BAM_SORT_STATS_SAMTOOLS ( 
            ch_alignment,
            ch_fasta
        )
        ch_stats    = BAM_SORT_STATS_SAMTOOLS.out.stats    // channel: [ val(meta), path(stats) ]
        ch_flagstat = BAM_SORT_STATS_SAMTOOLS.out.flagstat // channel: [ val(meta), path(flagstat) ]
        ch_idxstats = BAM_SORT_STATS_SAMTOOLS.out.idxstats // channel: [ val(meta), path(idxstats) ]
        ch_versions = ch_versions.mix(BAM_SORT_STATS_SAMTOOLS.out.versions)
    }
    else {
        FASTQ_ALIGN_BWA (
            ch_reads,
            ch_bwamem_index,
            true, // val_sort_bam hardcoded to true
            ch_fasta
        )
        ch_alignment        = ch_alignment.mix(FASTQ_ALIGN_BWA.out.bam)
        ch_alignment_index  = ch_alignment.mix(FASTQ_ALIGN_BWA.out.bai)
        ch_stats            = ch_alignment.mix(FASTQ_ALIGN_BWA.out.stats)    // channel: [ val(meta), path(stats) ]
        ch_flagstat         = ch_alignment.mix(FASTQ_ALIGN_BWA.out.flagstat) // channel: [ val(meta), path(flagstat) ]
        ch_idxstats         = ch_alignment.mix(FASTQ_ALIGN_BWA.out.idxstats) // channel: [ val(meta), path(idxstats) ]
        ch_versions         = ch_versions.mix(FASTQ_ALIGN_BWA.out.versions)
    }

    if (!skip_deduplication) {
        /*
         * Run Picard AddOrReplaceReadGroups to add read group (RG) to reads in bam file
         */

        PICARD_ADDORREPLACEREADGROUPS (
            ch_alignment,
            ch_fasta,
            ch_fasta_index
        )
        ch_versions = ch_versions.mix(PICARD_ADDORREPLACEREADGROUPS.out.versions)

        /*
         * Run Picard MarkDuplicates with the --REMOVE_DUPLICATES true flag
         */
        PICARD_REMOVEDUPLICATES (
            PICARD_ADDORREPLACEREADGROUPS.out.bam,
            ch_fasta,
            ch_fasta_index
        )
        ch_versions = ch_versions.mix(PICARD_REMOVEDUPLICATES.out.versions)

        /*
         * Run samtools index on deduplicated alignment
         */
        SAMTOOLS_INDEX_DEDUPLICATED (
            PICARD_REMOVEDUPLICATES.out.bam
        )
        ch_alignment       = PICARD_REMOVEDUPLICATES.out.bam
        ch_alignment_index = SAMTOOLS_INDEX_DEDUPLICATED.out.bai
        ch_picard_metrics  = PICARD_REMOVEDUPLICATES.out.metrics
        ch_versions        = ch_versions.mix(SAMTOOLS_INDEX_DEDUPLICATED.out.versions)
    }

    /*
     * Collect MultiQC inputs
     */
    ch_multiqc_files = ch_picard_metrics.collect{ meta, metrics -> metrics }
                        .mix(ch_flagstat.collect{ meta, flagstat -> flagstat })
                        .mix(ch_stats.collect{ meta, stats -> stats  })
                        .mix(ch_idxstats.collect{ meta, stats -> stats  })

    emit:
    bam                           = ch_alignment                     // channel: [ val(meta), [ bam ]       ]
    bai                           = ch_alignment_index               // channel: [ val(meta), [ bai ]       ]
    samtools_flagstat             = ch_flagstat                      // channel: [ val(meta), [ flagstat ]  ]
    samtools_stats                = ch_stats                         // channel: [ val(meta), [ stats ]     ]
    samtools_index_stats          = ch_idxstats                      // channel: [ val(meta), [ idxstats ]  ]
    picard_metrics                = ch_picard_metrics                // channel: [ val(meta), [ metrics ]   ]
    multiqc                       = ch_multiqc_files                 // channel: [ *{html,txt}              ]
    versions                      = ch_versions                      // channel: [ versions.yml             ]
}
