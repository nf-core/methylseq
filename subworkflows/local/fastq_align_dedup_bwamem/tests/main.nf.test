nextflow_workflow {

    name "Test Subworkflow FASTQ_ALIGN_DEDUP_BWAMEM"
    script "../main.nf"
    workflow "FASTQ_ALIGN_DEDUP_BWAMEM"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_align_dedup_bwamem"
    tag "bwa/mem"
    tag "parabricks/fq2bam"
    tag "samtools/sort"
    tag "samtools/index"
    tag "samtools/flagstat"
    tag "samtools/stats"
    tag "picard/markduplicates"
    tag "samtools/index"
    tag "untar"

    setup {
        run("UNTAR") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [:],
                    file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/Bwameth_Index.tar.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("Params: bwamem single-end - default") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                            [ id:'test', single_end:true ],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/SRR389222_sub1.fastq.gz', checkIfExists: true)
                ])
                input[1] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true)
                ])
                input[2] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa.fai', checkIfExists: true)
                ])
                input[3] = UNTAR.out.untar
                input[4] = false // skip_deduplication
                input[5] = false // use_gpu
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam.collect { meta, bamfile -> bam(bamfile).getReadsMD5() },
                    workflow.out.bai.collect { meta, bai -> file(bai).name },
                    workflow.out.samtools_flagstat,
                    workflow.out.samtools_stats,
                    workflow.out.samtools_index_stats,
                    workflow.out.picard_metrics.collect { meta, metrics -> file(metrics).name },
                    workflow.out.multiqc.flatten().collect { path -> file(path).name },
                    workflow.out.versions
                    ).match() }
            )
        }
    }

    test("Params: bwamem paired-end - default") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                            [ id:'test', single_end:true ],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/Ecoli_10K_methylated_R1.fastq.gz', checkIfExists: true),
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/Ecoli_10K_methylated_R2.fastq.gz', checkIfExists: true)
                ])
                input[1] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true)
                ])
                input[2] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa.fai', checkIfExists: true)
                ])
                input[3] = UNTAR.out.untar
                input[4] = false // skip_deduplication
                input[5] = false // use_gpu
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam.collect { meta, bamfile -> bam(bamfile).getReadsMD5() },
                    workflow.out.bai.collect { meta, bai -> file(bai).name },
                    workflow.out.samtools_flagstat,
                    workflow.out.samtools_stats,
                    workflow.out.samtools_index_stats,
                    workflow.out.picard_metrics.collect { meta, metrics -> file(metrics).name },
                    workflow.out.multiqc.flatten().collect { path -> file(path).name },
                    workflow.out.versions
                    ).match() }
            )
        }
    }

    test("Params: bwamem paired-end - skip_deduplication") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                            [ id:'test', single_end:true ],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/Ecoli_10K_methylated_R1.fastq.gz', checkIfExists: true),
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/testdata/Ecoli_10K_methylated_R2.fastq.gz', checkIfExists: true)
                ])
                input[1] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa', checkIfExists: true)
                ])
                input[2] = Channel.of([
                            [:],
                            file('https://github.com/nf-core/test-datasets/raw/methylseq/reference/genome.fa.fai', checkIfExists: true)
                ])
                input[3] = UNTAR.out.untar
                input[4] = true // skip_deduplication
                input[5] = false // use_gpu
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bam.collect { meta, bamfile -> bam(bamfile).getReadsMD5() },
                    workflow.out.bai.collect { meta, bai -> file(bai).name },
                    workflow.out.samtools_flagstat,
                    workflow.out.samtools_stats,
                    workflow.out.samtools_index_stats,
                    workflow.out.picard_metrics.collect { meta, metrics -> file(metrics).name },
                    workflow.out.multiqc.flatten().collect { path -> file(path).name },
                    workflow.out.versions
                    ).match() }
            )
        }
    }
}