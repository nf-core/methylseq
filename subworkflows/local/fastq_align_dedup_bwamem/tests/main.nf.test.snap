{
    "Params: bwamem single-end - default": {
        "content": [
            [
                "60a8b848db69b989825027aa9b57b6dc"
            ],
            [
                "test.dedup.sorted.bam.bai"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,fa1d26916675926472eca8792877bcc2"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,f8af98bf5fcab925dc5f6c22851622cb"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,f8af98bf5fcab925dc5f6c22851622cb"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,e4cc4f08ab947681aac06a51086b198c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,a0bbe1663eb64c8e429497ac947ae93b"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,f8af98bf5fcab925dc5f6c22851622cb"
                ]
            ],
            [
                "test.dedup.sorted.MarkDuplicates.metrics.txt"
            ],
            [
                "test.dedup.sorted.MarkDuplicates.metrics.txt",
                "test.sorted.bam",
                "test.flagstat",
                "test.sorted.bam",
                "test.idxstats",
                "test.sorted.bam",
                "test.stats"
            ],
            [
                "versions.yml:md5,25a1a4ccae61d15eca01af3e92d28ab5",
                "versions.yml:md5,4a350e432b223bcba51e65611dbdc632",
                "versions.yml:md5,4bd9092db8b15228c1fce19a449159d4",
                "versions.yml:md5,79cb0f0eebeecf8ea8379b23d0257e3b",
                "versions.yml:md5,846025f4bf158c6787e1993a75849e4b",
                "versions.yml:md5,86b45e50b28f949dac76b5bcc10460d5",
                "versions.yml:md5,8906f72a2cfba573aa0608980236503d",
                "versions.yml:md5,e3e4c46bf186965ef9260b603bfe834a",
                "versions.yml:md5,e74a07f6412edd9e98e4e4e4aacd15d0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-08-26T22:02:30.516109656"
    },
    "Params: bwamem paired-end - skip_deduplication": {
        "content": [
            [
                "5ec1463f3badcdb30b47e4a8f51c9978"
            ],
            [
                "test.sorted.bam",
                "test.sorted.bam.bai"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,523a01632ab25ab0445d0cf458303149"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,32c98b285a3e66ee6b02c7f1f10aab69"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,0a70bdf92cf4e6d99e9082622b102c53"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ]
            ],
            [

            ],
            [
                "test.sorted.bam",
                "test.flagstat",
                "test.sorted.bam",
                "test.idxstats",
                "test.sorted.bam",
                "test.stats"
            ],
            [
                "versions.yml:md5,25a1a4ccae61d15eca01af3e92d28ab5",
                "versions.yml:md5,4a350e432b223bcba51e65611dbdc632",
                "versions.yml:md5,79cb0f0eebeecf8ea8379b23d0257e3b",
                "versions.yml:md5,846025f4bf158c6787e1993a75849e4b",
                "versions.yml:md5,8906f72a2cfba573aa0608980236503d",
                "versions.yml:md5,e74a07f6412edd9e98e4e4e4aacd15d0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-08-26T22:03:23.540479755"
    },
    "Params: bwamem paired-end - default": {
        "content": [
            [
                "5ec1463f3badcdb30b47e4a8f51c9978"
            ],
            [
                "test.dedup.sorted.bam.bai"
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,523a01632ab25ab0445d0cf458303149"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,32c98b285a3e66ee6b02c7f1f10aab69"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,0a70bdf92cf4e6d99e9082622b102c53"
                ],
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.sorted.bam:md5,9b9fe8c083adda99cb579b8ed769321b"
                ]
            ],
            [
                "test.dedup.sorted.MarkDuplicates.metrics.txt"
            ],
            [
                "test.dedup.sorted.MarkDuplicates.metrics.txt",
                "test.sorted.bam",
                "test.flagstat",
                "test.sorted.bam",
                "test.idxstats",
                "test.sorted.bam",
                "test.stats"
            ],
            [
                "versions.yml:md5,25a1a4ccae61d15eca01af3e92d28ab5",
                "versions.yml:md5,4a350e432b223bcba51e65611dbdc632",
                "versions.yml:md5,4bd9092db8b15228c1fce19a449159d4",
                "versions.yml:md5,79cb0f0eebeecf8ea8379b23d0257e3b",
                "versions.yml:md5,846025f4bf158c6787e1993a75849e4b",
                "versions.yml:md5,86b45e50b28f949dac76b5bcc10460d5",
                "versions.yml:md5,8906f72a2cfba573aa0608980236503d",
                "versions.yml:md5,e3e4c46bf186965ef9260b603bfe834a",
                "versions.yml:md5,e74a07f6412edd9e98e4e4e4aacd15d0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-08-26T22:03:06.541388997"
    }
}
